Author: John Marshall <jm18@sanger.ac.uk>
Description: Add AC_SYS_LARGEFILE test 
 Defines _FILE_OFFSET_BITS in config.h as appropriate.  Hat tip @trifud.
 Now that config.h has ODR implications, add #include <config.h> to
 absolutely all *.c source files (cf 5f5aa02bfeb3857e8671f12a85787404ef9fa799).
 Most k*.c files differ from their klib upstream versions due to #include
 path-related trivia anyway, and knetfile.c in particular needs to see the
 newly-added definitions.
 .
 Picked from htslib's GitHub (f1327a703583bf61e2e7ffd11198487fee2a8422)
diff --git a/Makefile b/Makefile
index be2405e..1122fde 100644
--- a/Makefile
+++ b/Makefile
@@ -251,8 +251,8 @@ libhts.dylib: $(LIBHTS_OBJS)
 
 
 bgzf.o bgzf.pico: bgzf.c config.h $(htslib_hts_h) $(htslib_bgzf_h) $(htslib_hfile_h) $(htslib_khash_h)
-kstring.o kstring.pico: kstring.c $(htslib_kstring_h)
-knetfile.o knetfile.pico: knetfile.c $(htslib_knetfile_h)
+kstring.o kstring.pico: kstring.c config.h $(htslib_kstring_h)
+knetfile.o knetfile.pico: knetfile.c config.h $(htslib_knetfile_h)
 hfile.o hfile.pico: hfile.c config.h $(htslib_hfile_h) $(hfile_internal_h) $(hts_internal_h) $(htslib_khash_h)
 hfile_irods.o hfile_irods.pico: hfile_irods.c config.h $(hfile_internal_h) $(htslib_hts_h) $(htslib_kstring_h)
 hfile_libcurl.o hfile_libcurl.pico: hfile_libcurl.c config.h $(hfile_internal_h) $(htslib_hts_h) $(htslib_kstring_h)
@@ -265,7 +265,7 @@ faidx.o faidx.pico: faidx.c config.h $(htslib_bgzf_h) $(htslib_faidx_h) $(htslib
 synced_bcf_reader.o synced_bcf_reader.pico: synced_bcf_reader.c config.h $(htslib_synced_bcf_reader_h) $(htslib_kseq_h) $(htslib_khash_str2int_h) $(htslib_bgzf_h)
 vcf_sweep.o vcf_sweep.pico: vcf_sweep.c config.h $(htslib_vcf_sweep_h) $(htslib_bgzf_h)
 vcfutils.o vcfutils.pico: vcfutils.c config.h $(htslib_vcfutils_h) $(htslib_kbitset_h)
-kfunc.o kfunc.pico: kfunc.c $(htslib_kfunc_h)
+kfunc.o kfunc.pico: kfunc.c config.h $(htslib_kfunc_h)
 regidx.o regidx.pico: regidx.c config.h $(htslib_hts_h) $(htslib_kstring_h) $(htslib_kseq_h) $(htslib_khash_str2int_h) $(htslib_regidx_h)
 md5.o md5.pico: md5.c config.h $(htslib_hts_h)
 plugin.o plugin.pico: plugin.c config.h $(hts_internal_h) $(htslib_kstring_h)
@@ -335,13 +335,13 @@ test/test-vcf-api: test/test-vcf-api.o libhts.a
 test/test-vcf-sweep: test/test-vcf-sweep.o libhts.a
 	$(CC) -pthread $(LDFLAGS) -o $@ test/test-vcf-sweep.o libhts.a -lz $(LIBS)
 
-test/fieldarith.o: test/fieldarith.c $(htslib_sam_h)
-test/hfile.o: test/hfile.c $(htslib_hfile_h) $(htslib_hts_defs_h)
-test/sam.o: test/sam.c $(htslib_sam_h) $(htslib_faidx_h) $(htslib_kstring_h)
-test/test-regidx.o: test/test-regidx.c $(htslib_regidx_h)
-test/test_view.o: test/test_view.c $(cram_h) $(htslib_sam_h)
-test/test-vcf-api.o: test/test-vcf-api.c $(htslib_hts_h) $(htslib_vcf_h) $(htslib_kstring_h) $(htslib_kseq_h)
-test/test-vcf-sweep.o: test/test-vcf-sweep.c $(htslib_vcf_sweep_h)
+test/fieldarith.o: test/fieldarith.c config.h $(htslib_sam_h)
+test/hfile.o: test/hfile.c config.h $(htslib_hfile_h) $(htslib_hts_defs_h)
+test/sam.o: test/sam.c config.h $(htslib_sam_h) $(htslib_faidx_h) $(htslib_kstring_h)
+test/test-regidx.o: test/test-regidx.c config.h $(htslib_regidx_h)
+test/test_view.o: test/test_view.c config.h $(cram_h) $(htslib_sam_h)
+test/test-vcf-api.o: test/test-vcf-api.c config.h $(htslib_hts_h) $(htslib_vcf_h) $(htslib_kstring_h) $(htslib_kseq_h)
+test/test-vcf-sweep.o: test/test-vcf-sweep.c config.h $(htslib_vcf_sweep_h)
 
 
 install: libhts.a $(BUILT_PROGRAMS) $(BUILT_PLUGINS) installdirs install-$(SHLIB_FLAVOUR) install-pkgconfig
diff --git a/config.h.in b/config.h.in
index 09b23ff..0397762 100644
--- a/config.h.in
+++ b/config.h.in
@@ -25,3 +25,14 @@
 
 /* Define to 1 if you have a working `mmap' system call. */
 #undef HAVE_MMAP
+
+/* Enable large inode numbers on Mac OS X 10.5.  */
+#ifndef _DARWIN_USE_64_BIT_INODE
+# define _DARWIN_USE_64_BIT_INODE 1
+#endif
+
+/* Number of bits in a file offset, on hosts where this is settable. */
+#undef _FILE_OFFSET_BITS
+
+/* Define for large files, on AIX-style hosts. */
+#undef _LARGE_FILES
diff --git a/configure.ac b/configure.ac
index 27fa1f4..6f658a2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -38,6 +38,8 @@ redistribute it.  There is NO WARRANTY, to the extent permitted by law.])
 AC_PROG_CC
 AC_PROG_RANLIB
 
+AC_SYS_LARGEFILE
+
 need_crypto=no
 
 AC_ARG_WITH([irods],
diff --git a/kfunc.c b/kfunc.c
index faa2ac5..323e70f 100644
--- a/kfunc.c
+++ b/kfunc.c
@@ -24,6 +24,8 @@
    SOFTWARE.
 */
 
+#include <config.h>
+
 #include <math.h>
 #include <stdlib.h>
 #include "htslib/kfunc.h"
diff --git a/knetfile.c b/knetfile.c
index 28fe629..aa28858 100644
--- a/knetfile.c
+++ b/knetfile.c
@@ -28,6 +28,8 @@
    therefore I decide to heavily annotate this file, for Linux and
    Windows as well.  -ac */
 
+#include <config.h>
+
 #include <time.h>
 #include <stdio.h>
 #include <ctype.h>
diff --git a/kstring.c b/kstring.c
index b84ee2c..f43a982 100644
--- a/kstring.c
+++ b/kstring.c
@@ -23,6 +23,8 @@
    SOFTWARE.
 */
 
+#include <config.h>
+
 #include <stdarg.h>
 #include <stdio.h>
 #include <ctype.h>
diff --git a/test/fieldarith.c b/test/fieldarith.c
index 6d871b7..3627a89 100644
--- a/test/fieldarith.c
+++ b/test/fieldarith.c
@@ -22,6 +22,8 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.  */
 
+#include <config.h>
+
 #include <stdio.h>
 
 #include "htslib/sam.h"
diff --git a/test/hfile.c b/test/hfile.c
index c4ba91c..180f2ff 100644
--- a/test/hfile.c
+++ b/test/hfile.c
@@ -22,6 +22,8 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.  */
 
+#include <config.h>
+
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
diff --git a/test/sam.c b/test/sam.c
index d6610c0..2c85010 100644
--- a/test/sam.c
+++ b/test/sam.c
@@ -22,6 +22,8 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.  */
 
+#include <config.h>
+
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
diff --git a/test/test-regidx.c b/test/test-regidx.c
index 0aea6b8..abb390b 100644
--- a/test/test-regidx.c
+++ b/test/test-regidx.c
@@ -23,6 +23,8 @@
     THE SOFTWARE.
 */
 
+#include <config.h>
+
 #include <stdarg.h>
 #include <stdlib.h>
 #include <stdio.h>
diff --git a/test/test-vcf-api.c b/test/test-vcf-api.c
index 3e7623a..2df9b71 100644
--- a/test/test-vcf-api.c
+++ b/test/test-vcf-api.c
@@ -22,6 +22,8 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.  */
 
+#include <config.h>
+
 #include <stdio.h>
 #include <htslib/hts.h>
 #include <htslib/vcf.h>
diff --git a/test/test-vcf-sweep.c b/test/test-vcf-sweep.c
index 57c47bc..44c2fb1 100644
--- a/test/test-vcf-sweep.c
+++ b/test/test-vcf-sweep.c
@@ -22,6 +22,8 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.  */
 
+#include <config.h>
+
 #include <stdio.h>
 #include <htslib/vcf_sweep.h>
 
diff --git a/test/test_view.c b/test/test_view.c
index 6dbef00..0b83bf5 100644
--- a/test/test_view.c
+++ b/test/test_view.c
@@ -23,6 +23,8 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.  */
 
+#include <config.h>
+
 #include <stdio.h>
 #include <unistd.h>
 #include <stdlib.h>
